### ISSUE TITLE
Partials are not properly taken into account when finding cycles

### ISSUE THREAD

**Issue by ZyX-I** [vim/vim#731]
Consider the following code:

```
let l = []
let Pt = function('string', [l])
call add(l, Pt)
unlet l
unlet Pt
call garbagecollect(1)
```

In this case partial Pt is referenced inside l, so `unlet Pt` will not free it. `unlet l` will not free l because it is referenced in Pt. But when searching for cycles it will find “unreferenced” list l which contains reference to Pt which contains reference to l, so as the result you will see double free. Valgrind report for

```
valgrind --leak-check=full --read-var-info=yes --track-origins=yes --num-callers=50 --trace-children=no --log-file=valgrind.log --tool=memcheck ../vim -u NONE -i NONE -N -c 'let l=[]' -c 'let Pt=function("string", [l])' -c 'call add(l, Pt)' -c 'unlet l' -c 'unlet Pt' -c 'call garbagecollect(1)'
:qa<CR>
```

below

```
==15174== Memcheck, a memory error detector
==15174== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==15174== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==15174== Command: ../vim -u NONE -i NONE -N -c let\ l=[] -c let\ Pt=function("string",\ [l]) -c call\ add(l,\ Pt) -c unlet\ l -c unlet\ Pt -c call\ garbagecollect(1)
==15174== Parent PID: 21419
==15174== 

parse DIE(readdwarf3.c:3651): confused by:
 <2><268b>: Abbrev Number: 26 (DW_TAG_subrange_type)
     DW_AT_type        : <2692> 
     DW_AT_count       : 64 
parse_type_DIE:
--15174-- WARNING: Serious error when reading debug info
--15174-- When reading debug info from /home/zyx/a.a/Proj/c/vim/src/vim:
--15174-- confused by the above DIE
==15174== Invalid read of size 8
==15174==    at 0x483CE8: list_free (eval.c:6044)
==15174==    by 0x484CC8: free_unref_items (eval.c:7022)
==15174==    by 0x47AC03: garbage_collect (eval.c:6950)
==15174==    by 0x515431: vgetc (getchar.c:1574)
==15174==    by 0x517F4C: safe_vgetc (getchar.c:1801)
==15174==    by 0x57AA09: normal_cmd (normal.c:627)
==15174==    by 0x6EF55B: main_loop (main.c:1353)
==15174==    by 0x6EB85C: main (main.c:1045)
==15174==  Address 0x1059f330 is 0 bytes inside a block of size 88 free'd
==15174==    at 0x4C2B470: free (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==15174==    by 0x564AD3: vim_free (misc2.c:1695)
==15174==    by 0x483CFF: list_free (eval.c:6053)
==15174==    by 0x47A633: list_unref (eval.c:6022)
==15174==    by 0x47D7DA: clear_tv (eval.c:21955)
==15174==    by 0x484FAD: partial_free (eval.c:7199)
==15174==    by 0x484F61: partial_unref (eval.c:7216)
==15174==    by 0x47D7BC: clear_tv (eval.c:21951)
==15174==    by 0x483CD7: list_free (eval.c:6050)
==15174==    by 0x484CC8: free_unref_items (eval.c:7022)
==15174==    by 0x47AC03: garbage_collect (eval.c:6950)
==15174==    by 0x515431: vgetc (getchar.c:1574)
==15174==    by 0x517F4C: safe_vgetc (getchar.c:1801)
==15174==    by 0x57AA09: normal_cmd (normal.c:627)
==15174==    by 0x6EF55B: main_loop (main.c:1353)
==15174==    by 0x6EB85C: main (main.c:1045)
==15174==  Block was alloc'd at
==15174==    at 0x4C2A0B0: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==15174==    by 0x564320: lalloc (misc2.c:920)
==15174==    by 0x5644C7: alloc_clear (misc2.c:842)
==15174==    by 0x47A311: list_alloc (eval.c:5983)
==15174==    by 0x4AE86C: get_list_tv (eval.c:5921)
==15174==    by 0x4AD8B2: eval7 (eval.c:5162)
==15174==    by 0x4AD0CE: eval6 (eval.c:4889)
==15174==    by 0x4ACC44: eval5 (eval.c:4705)
==15174==    by 0x4ABEEA: eval4 (eval.c:4412)
==15174==    by 0x4ABCC9: eval3 (eval.c:4327)
==15174==    by 0x4ABAB9: eval2 (eval.c:4259)
==15174==    by 0x47D9E2: eval1 (eval.c:4187)
==15174==    by 0x47D4B3: eval0 (eval.c:4147)
==15174==    by 0x47F640: ex_let (eval.c:1956)
==15174==    by 0x4D12F6: do_one_cmd (ex_docmd.c:2921)
==15174==    by 0x4CCB0D: do_cmdline (ex_docmd.c:1107)
==15174==    by 0x4CD9B5: do_cmdline_cmd (ex_docmd.c:712)
==15174==    by 0x6EEBC7: exe_commands (main.c:2927)
==15174==    by 0x6EB65D: main (main.c:955)
==15174== 
==15174== Invalid free() / delete / delete[] / realloc()
==15174==    at 0x4C2B470: free (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==15174==    by 0x564AD3: vim_free (misc2.c:1695)
==15174==    by 0x483CFF: list_free (eval.c:6053)
==15174==    by 0x484CC8: free_unref_items (eval.c:7022)
==15174==    by 0x47AC03: garbage_collect (eval.c:6950)
==15174==    by 0x515431: vgetc (getchar.c:1574)
==15174==    by 0x517F4C: safe_vgetc (getchar.c:1801)
==15174==    by 0x57AA09: normal_cmd (normal.c:627)
==15174==    by 0x6EF55B: main_loop (main.c:1353)
==15174==    by 0x6EB85C: main (main.c:1045)
==15174==  Address 0x1059f330 is 0 bytes inside a block of size 88 free'd
==15174==    at 0x4C2B470: free (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==15174==    by 0x564AD3: vim_free (misc2.c:1695)
==15174==    by 0x483CFF: list_free (eval.c:6053)
==15174==    by 0x47A633: list_unref (eval.c:6022)
==15174==    by 0x47D7DA: clear_tv (eval.c:21955)
==15174==    by 0x484FAD: partial_free (eval.c:7199)
==15174==    by 0x484F61: partial_unref (eval.c:7216)
==15174==    by 0x47D7BC: clear_tv (eval.c:21951)
==15174==    by 0x483CD7: list_free (eval.c:6050)
==15174==    by 0x484CC8: free_unref_items (eval.c:7022)
==15174==    by 0x47AC03: garbage_collect (eval.c:6950)
==15174==    by 0x515431: vgetc (getchar.c:1574)
==15174==    by 0x517F4C: safe_vgetc (getchar.c:1801)
==15174==    by 0x57AA09: normal_cmd (normal.c:627)
==15174==    by 0x6EF55B: main_loop (main.c:1353)
==15174==    by 0x6EB85C: main (main.c:1045)
==15174==  Block was alloc'd at
==15174==    at 0x4C2A0B0: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==15174==    by 0x564320: lalloc (misc2.c:920)
==15174==    by 0x5644C7: alloc_clear (misc2.c:842)
==15174==    by 0x47A311: list_alloc (eval.c:5983)
==15174==    by 0x4AE86C: get_list_tv (eval.c:5921)
==15174==    by 0x4AD8B2: eval7 (eval.c:5162)
==15174==    by 0x4AD0CE: eval6 (eval.c:4889)
==15174==    by 0x4ACC44: eval5 (eval.c:4705)
==15174==    by 0x4ABEEA: eval4 (eval.c:4412)
==15174==    by 0x4ABCC9: eval3 (eval.c:4327)
==15174==    by 0x4ABAB9: eval2 (eval.c:4259)
==15174==    by 0x47D9E2: eval1 (eval.c:4187)
==15174==    by 0x47D4B3: eval0 (eval.c:4147)
==15174==    by 0x47F640: ex_let (eval.c:1956)
==15174==    by 0x4D12F6: do_one_cmd (ex_docmd.c:2921)
==15174==    by 0x4CCB0D: do_cmdline (ex_docmd.c:1107)
==15174==    by 0x4CD9B5: do_cmdline_cmd (ex_docmd.c:712)
==15174==    by 0x6EEBC7: exe_commands (main.c:2927)
==15174==    by 0x6EB65D: main (main.c:955)
==15174== 
==15174== 
<leak report skipped>
==15174== 
==15174== For counts of detected and suppressed errors, rerun with: -v
==15174== ERROR SUMMARY: 23 errors from 23 contexts (suppressed: 0 from 0)
```

This is vim-7.4.1646 (with my patch to make vim.Function create a partial, but it did not touch the relevant code), but AFAIK there were no fixes. One needs to treat partial like any other container. I accidentally hit the similar bug when creating my vim.Function tests: partials must be stored in a linked list like any other container and when freeing lists, dictionaries or partials with second argument `FALSE` lists dictionaries and partials must not be freed, not only lists and dictionaries; same in partial_free where second argument needs to be `recurse` as well. Currently only `dict_free` has some special code for partials, but it is _not_ a correct code because it only handles `pt_dict` specially:

```
let d = {}
let Pt = function('string', [d])
let d.Pt = Pt
unlet d
unlet Pt
call garbagecollect(1)
```

yields absolutely the same bug.

**Comment by brammool** [comment]
Nikolai Pavlov wrote:

> Consider the following code:
> 
> ```
> let l = []
> let Pt = function('string', [l])
> call add(l, Pt)
> unlet l
> unlet Pt
> call garbagecollect(1)
> ```
> 
> In this case partial Pt is referenced inside l, so `unlet Pt` will not free it. `unlet l` will not free l because it is referenced in Pt. But when searching for cycles it will find “unreferenced” list l which contains reference to Pt which contains reference to l, so as the result you will see double free. Valgrind report for
> 
> ```
> valgrind --leak-check=full --read-var-info=yes --track-origins=yes --num-callers=50 --trace-children=no --log-file=valgrind.log --tool=memcheck ../vim -u NONE -i NONE -N -c 'let l=[]' -c 'let Pt=function("string", [l])' -c 'call add(l, Pt)' -c 'unlet l' -c 'unlet Pt' -c 'call garbagecollect(1)'
> :qa<CR>
> ```
> 
> below

Thanks for reporting.

> This is vim-7.4.1646 (with my patch to make vim.Function create a
> partial, but it did not touch the relevant code), but AFAIK there were
> no fixes. One needs to treat partial like any other container. I
> accidentally hit the similar bug when creating my vim.Function tests:
> partials must be stored in a linked list like any other container and
> when freeing lists, dictionaries or partials with second argument
> `FALSE` lists dictionaries and partials must not be freed, not only
> lists and dictionaries; same in partial_free where second argument
> needs to be `recurse` as well. Currently only `dict_free` has some
> special code for partials, but it is _not_ a correct code because it
> only handles `pt_dict` specially:
> 
> ```
> let d = {}
> let Pt = function('string', [d])
> let d.Pt = Pt
> unlet d
> unlet Pt
> call garbagecollect(1)
> ```
> 
> yields absolutely the same bug.

I thought I fixed this, but apparently there is still something wrong.

## 

hundred-and-one symptoms of being an internet addict:
235. You start naming your kids Pascal, COBOL, Algol and Fortran.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
closed by 7.4.1715
